experiment/build/generate_cloudbuild.py

```diff
@@ -13,65 +13,119 @@
+    tag += ':' + experiment_utils.get_experiment_name()
+    return tag
+
+
+def coverage_steps(benchmark):
+    """Returns GCB run steps for coverage builds."""
+    coverage_binaries_dir = exp_path.filestore(
+        build_utils.get_coverage_binaries_dir())
+    steps = [{
+        'name':
+            'gcr.io/cloud-builders/docker',
+        'args': [
+            'run', '-v', '/workspace/out:/host-out',
+            posixpath.join(experiment_utils.get_base_docker_tag(), 'builders',
+                           'coverage', benchmark) + ':' +
+            experiment_utils.get_experiment_name(), '/bin/bash', '-c',
+            'cd /out; tar -czvf /host-out/coverage-build-' + benchmark +
+            '.tar.gz *'
+        ]
+    }]
+    step = {'name': 'gcr.io/cloud-builders/gsutil'}
+    step['args'] = [
+        '-m', 'cp', '/workspace/out/coverage-build-' + benchmark + '.tar.gz',
+        coverage_binaries_dir + '/'
+    ]
+    steps.append(step)
+    return steps
+
+
+def create_cloud_build_spec(image_templates,
+                            benchmark='',
+                            experiment=True,
+                            build_base_images=False):
+    """Generates Cloud Build specification.
+
+    Args:
+      image_templates: Image types and their properties.
+      benchmark: Name of benchmark (required for coverage builds only).
+      build_base_images: True if building only base images.
+
+    Returns:
+      GCB build steps.
+    """
+    cloud_build_spec = {'steps': [], 'images': []}
+
+    for image_name, image_specs in image_templates.items():
+        step = {
+            'id': image_name,
+            'env': 'DOCKER_BUILDKIT=1',
+            'name': 'gcr.io/cloud-builders/docker'
+        }
+        step['args'] = [
+            'build', '--tag',
+            posixpath.join(BASE_TAG, image_specs['tag']), '--tag',
+            get_experiment_tag_for_image(image_specs,
+                                         experiment), '--cache-from',
+            get_experiment_tag_for_image(image_specs, experiment).split(':')[0],
+            '--build-arg', 'BUILDKIT_INLINE_CACHE=1'
+        ]
+        for build_arg in image_specs.get('build_arg', []):
+            step['args'] += ['--build-arg', build_arg]
+
+        step['args'] += [
+            '--file', image_specs['dockerfile'], image_specs['path']
+        ]
+        step['wait_for'] = []
+        for dep in image_specs.get('depends_on', []):
+            # Base images are built before creating fuzzer benchmark builds,
+            # so it's not required to wait for them to build.
+            if 'base' in dep and not build_base_images:
+                continue
+            step['wait_for'] += [dep]
+
         cloud_build_spec['steps'].append(step)
+        cloud_build_spec['images'].append(
+            get_experiment_tag_for_image(image_specs, experiment))
+        cloud_build_spec['images'].append(
```

Comment:
I think it would be nicer if get_experiment_tag_for_image returned multiple tags instead of parsing the result and creating another one. Is that posisble?

Reply:
Yes, would use a flag.